api/controller.go

```diff
@@ -2372,6 +2396,68 @@ func (c *Controller) DetachPolicyFromGroupHandler() authop.DetachPolicyFromGroup
 	})
 }
 
+func (c *Controller) RefsRefsRestoreHandler() refs.RefsRestoreHandler {
+	return refs.RefsRestoreHandlerFunc(func(params refs.RefsRestoreParams, user *models.User) middleware.Responder {
+		deps, err := c.setupRequest(user, params.HTTPRequest, []permissions.Permission{
+			{
+				Action:   permissions.ListTagsAction,
+				Resource: permissions.RepoArn(params.Repository),
+			},
+			{
+				Action:   permissions.ListBranchesAction,
+				Resource: permissions.RepoArn(params.Repository),
+			},
+			{
+				Action:   permissions.ListCommitsAction,
+				Resource: permissions.RepoArn(params.Repository),
+			},
+		})
+		if err != nil {
+			return refs.NewRefsRestoreUnauthorized().
+				WithPayload(responseErrorFrom(err))
+		}
+		deps.LogAction("restore_repository_refs")
+
+		repo, err := deps.Cataloger.GetRepository(deps.ctx, params.Repository)
+		if errors.Is(err, catalog.ErrRepositoryNotFound) {
+			return refs.NewRefsRestoreNotFound().
+				WithPayload(responseErrorFrom(err))
+		} else if err != nil {
+			return refs.NewRefsRestoreDefault(http.StatusInternalServerError).
+				WithPayload(responseErrorFrom(err))
+		}
+
+		// ensure no refs currently found
+		_, _, err = deps.Cataloger.ListCommits(deps.ctx, repo.Name, repo.DefaultBranch, "", 1)
+		if !errors.Is(err, graveler.ErrNotFound) {
+			return refs.NewRefsRestoreDefault(http.StatusInternalServerError).
```

Comment:
Consider reporting more specific (non-internal) errors for business logic failures.

Reply:
Actually you caught a bug, it was meant to be a `NewRestoreBadRequest` instead.